crates/rustc_codegen_spirv/src/codegen_cx/entry.rs

```diff
@@ -108,124 +113,136 @@ impl<'tcx> CodegenCx<'tcx> {
-        let fn_id = emit
-            .begin_function(void, None, FunctionControl::NONE, fn_void_void)
-            .unwrap();
-        emit.begin_block(None).unwrap();
-        // Adjust any global `OpVariable`s as needed (e.g. loading from `Input`s).
-        let arguments: Vec<_> = declared_params
+        let mut bx = Builder::new_block(self, stub_fn, "");
+        // Adjust any global `OpVariable`s as needed (e.g. loading from `Input`s,
+        // or accessing the sole field of an "interface block" `OpTypeStruct`),
+        // to match the argument type we have to pass to the Rust entry `fn`.
+        let arguments: Vec<_> = interface_globals
             .iter()
             .zip(arg_abis)
             .zip(hir_params)
-            .flat_map(|((&(var, storage_class), entry_fn_arg), hir_param)| {
-                let mut dst_len_arg = None;
-                let arg = match entry_fn_arg.layout.ty.kind() {
-                    TyKind::Ref(_, ty, _) => {
-                        if !ty.is_sized(self.tcx.at(span), self.param_env()) {
-                            dst_len_arg.replace(
-                                self.dst_length_argument(&mut emit, ty, hir_param, len_t, var),
-                            );
+            .flat_map(
+                |((&(global_var, storage_class), entry_fn_arg), hir_param)| {
+                    bx.set_span(hir_param.span);
+
+                    let var_value_spirv_type = match self.lookup_type(global_var.ty) {
+                        SpirvType::Pointer { pointee } => pointee,
+                        _ => unreachable!(),
+                    };
+
+                    let (first, second) = match entry_fn_arg.layout.ty.kind() {
+                        TyKind::Ref(_, pointee_ty, _) => {
+                            let arg_pointee_spirv_type = self
+                                .layout_of(pointee_ty)
+                                .spirv_type(hir_param.ty_span, self);
+
+                            if let SpirvType::InterfaceBlock { inner_type } =
+                                self.lookup_type(var_value_spirv_type)
+                            {
+                                assert_ty_eq!(self, arg_pointee_spirv_type, inner_type);
+
+                                let inner = bx.struct_gep(global_var, 0);
+
+                                match entry_fn_arg.mode {
+                                    PassMode::Direct(_) => (inner, None),
+
+                                    // Unsized pointee with length (i.e. `&[T]`).
+                                    PassMode::Pair(..) => {
+                                        // FIXME(eddyb) shouldn't this be `usize`?
+                                        let len_spirv_type = self.type_isize();
+
+                                        let len = bx
+                                            .emit()
+                                            .array_length(
+                                                len_spirv_type,
```

Comment:
12 levels of indentation makes me sad :( would be nice to split this out

Reply:
I guess the newly introduce use of `Builder` means that we can interleave things a bit more and not have to separate them out so much - e.g. `declare_interface_global_for_param` could also return an `OperandValue` for the call args.